### PR TITLE
Manifest-driven shed repository definitions

### PR DESCRIPTION
**Pull request by jmchilton**
Currently there exists a tension between what is best for developers (storing all tools in a single repository - e.g. [ncbi_blast_plus](https://github.com/peterjc/galaxy_blast/tree/master/tools/ncbi_blast_plus) or [bedtools](https://github.com/galaxyproject/tools-iuc/tree/master/tools/bedtools)) and what is best for Galaxy users (storing a single repository per tool and collecting them together with a suite - e.g. [samtools](https://github.com/galaxyproject/tools-devteam/tree/master/tool_collections/samtools) or [gatk](https://github.com/galaxyproject/tools-devteam/tree/master/tool_collections/gatk)). More discussion [here](https://github.com/galaxyproject/planemo/issues/131#issuecomment-94817086).

This pull request extends the semantics of ``.shed.yml`` in a attempt to resolve this tension and make the best practice for Galaxy users trivial to manage for developers. Previously each ``.shed.yml`` could only correspond to a single Tool Shed repository and it would collect all files in a directory (except an optional list of [ignored files](https://github.com/galaxyproject/planemo/pull/99)). This pull request extends the ``shed_create`` and ``shed_upload`` commands to allow ``.shed.yml`` files to correspond to any number of actual Tool Shed repositories each with fully customizable file includes and excludes.

While there is a great deal customization allowed - two new keys ``auto_tool_repositories`` and ``suite`` provide shortcuts to quickly and implicity define repositories for for each individual tool in the directory and build a suite for those. Consider the following (admittedly idealized) samtools example:

```
owner: "devteam"
remote_repository_url: "https://github.com/galaxyproject/tools-devteam/tool_collections/samtools"
homepage_url: "https://github.com/galaxyproject/tools-devteam/"
categories:
  - "SAM"
auto_tool_repositories:
  name_template: "{{ tool_id }}"
  description_template: "Wrapper for samtools application {{ tool_name }}."
suite:
  name: "suite_samtools_1_2"
  description: "A suite of Galaxy tools designed to work with version 1.2 of the SAMtools package."
  long_description:
  > SAM (Sequence Alignment/Map) format is a generic format for storing large nucleotide sequence
    alignments.   This repository suite associates selected repositories containing Galaxy utilities that require
    version 1.2 of the SAMTools package.  These associated Galaxy utilities consist of a Galaxy Data
    Manager contained in the repository named data_manager_sam_fasta_index_builder and Galaxy tools
    contained in several separate repositories.
```

This example assumes the ``.shed.yml`` file is placed in a "flat" directory with each samtools tool wrapper and planemo will create and update repositories for each individual tool given the specified templates in ``auto_tool_repositories``. The ``suite`` key here will auto-generate a suite repository for all of these tools and will automatically created the corresponding ``repository_dependencies.xml`` to populate it with (this is generated during ``shed_upload`` and never needs to exist in your repository).

Again this example is admittedly idealized, but if ``auto_tool_repositories`` is not specified, a ``repositories`` list can be specified instead. There are some examples of this in the test data included with this pull request:

- [This .shed.yml](https://github.com/jmchilton/planemo/blob/shed_realizations/tests/data/repos/multi_repos_flat_configured/.shed.yml) is a simple example of specifying custom repositories for individual tools.
-[This](https://github.com/jmchilton/planemo/blob/shed_realizations/tests/data/repos/multi_repos_flat_configured_complex/.shed.yml) demonstrates complex inclusions files from sub-directories and renaming.
- [This .shed.yml](https://github.com/jmchilton/planemo/blob/shed_realizations/tests/data/repos/multi_repos_flat_configured_complex/.shed.yml) demonstrates complex inclusions files from sub-directories and renaming.

The test data also includes some more advanced usages of the ``suite`` key as well - specifically using it [without auto_tool_repositories](https://github.com/jmchilton/planemo/blob/shed_realizations/tests/data/repos/suite_auto/.shed.yml) as a generic replacement for ``repository_dependencies.xml`` and adding [additional dependent repositories](https://github.com/jmchilton/planemo/blob/shed_realizations/tests/data/repos/multi_repos_flat_flag_suite/.shed.yml) in addition to the ones defined by the ``.shed.yml`` file.

Implements #26.

### CODE REVIEW

File: planemo/shed.py [hunk 1]
```diff
@@ -193,23 +200,132 @@ def _find_repository_id(ctx, tsi, name, repo_config, **kwds):
...
+        "include": [],
+        "name": name,
+        "description": description,
+    }
+    if long_description:
+        repo["long_description"] = long_description
+    repos[name] = repo
+
+
+def find_repository(tsi, owner, name):
```
Comment (erasche) [inline planemo/shed.py]: Does this belong in shed util?

Comment (jmchilton) [inline planemo/shed.py]: What is shed util?

Comment (erasche) [inline planemo/shed.py]: Oh, I meant the shed.py file in planemo where a bunch of TS interactivity
was stuck

man. 27. apr. 2015, 08.32 skrev John Chilton <notifications@github.com>:

Comment (jmchilton) [inline planemo/shed.py]: That is this file :).

**Comment by jmchilton** [comment]
I would like to do a planemo release tonight - preferably including this pull request. Let me know if this ``.shed.yml`` additions in here rub anyone the wrong way and I will do the release without this.

File: planemo/shed.py (hunk 1 above)
Comment (erasche) [inline planemo/shed.py]: I...uh...how could I possibly be this blind. So sorry @jm

**Comment by erasche** [comment]
+1

**Comment by erasche** [comment]
https://github.com/galaxy-iuc/standards/issues/6

**Comment by jmchilton** [comment]
Awesome - thanks for the review @erasche.

**Pull request merged**